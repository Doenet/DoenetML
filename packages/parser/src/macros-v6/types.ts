export interface ParseOptions {
    filename?: string;
    startRule?: "start";
    tracer?: any;
    [key: string]: any;
}
export type ParseFunction = <Options extends ParseOptions>(
    input: string,
    options?: Options,
) => Options extends { startRule: infer StartRule }
    ? StartRule extends "start"
        ? Top
        : Top
    : Top;

// These types were autogenerated by ts-pegjs
export type Top = (Macro | FunctionMacro | Text)[];
export type Ident = string;
export type ScopedIdent = string;
export type PathPart = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "pathPart"; name: Ident; index: PropIndex[] };
export type ScopedPathPart = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "pathPart"; name: ScopedIdent; index: PropIndex[] };
export type EmptyPathPart = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "pathPart"; name: ""; index: [] };
export type FullPath =
  | [EmptyPathPart, ...ScopedPathPartOrDD[]]
  | [ScopedPathPartOrDD, ...ScopedPathPartOrDD[]];
export type ScopedPathPartOrDD =
  | ScopedPathPart
  | ({
      position: {
        start: { offset: number; line: number; column: number };
        end: { offset: number; line: number; column: number };
      };
    } & { type: "pathPart"; name: ".."; index: [] });
export type PropAccess = PartialPathMacro;
export type ScopedPropAccess = ScopedPartialPathMacro;
export type Macro = FullPathMacro | PartialPathMacro;
export type PartialPathMacro = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & {
  type: "macro";
  version: "0.6";
  path: [PathPart];
  attributes: never[] | NonNullable<PropAttrs | null>;
  accessedProp: PropAccess | null;
};
export type ScopedPartialPathMacro = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & {
  type: "macro";
  version: "0.6";
  path: [ScopedPathPart];
  attributes: never[] | NonNullable<PropAttrs | null>;
  accessedProp: ScopedPropAccess | null;
};
export type FullPathMacro = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & {
  type: "macro";
  version: "0.6";
  path: FullPath;
  attributes: never[] | NonNullable<PropAttrs | null>;
  accessedProp: ScopedPropAccess | null;
};
export type FunctionMacro =
  | ({
      position: {
        start: { offset: number; line: number; column: number };
        end: { offset: number; line: number; column: number };
      };
    } & { type: "function"; version: "0.6"; macro: FullPathMacro; input: FunctionInput | null })
  | ({
      position: {
        start: { offset: number; line: number; column: number };
        end: { offset: number; line: number; column: number };
      };
    } & {
      type: "function";
      macro: PartialPathMacro;
      version: "0.6";
      input: FunctionInput | null;
    });
export type FunctionInput = FunctionArgumentList;
export type FunctionArgumentList = [
  BalancedParenTextNoComma,
  ...BalancedParenTextNoComma[]
];
export type BalancedParenTextNoComma = (Macro | FunctionMacro | Text)[];
export type BalancedParenText = (Macro | FunctionMacro | Text)[];
export type PropAttrs = Record<string, Attr>;
export type PropIndex = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "index"; value: (FunctionMacro | Macro | TextWithoutClosingSquareBrace)[] };
export type Attr =
  | ({
      position: {
        start: { offset: number; line: number; column: number };
        end: { offset: number; line: number; column: number };
      };
    } & { type: "attribute"; name: AttrName; children: AttrValue })
  | ({
      position: {
        start: { offset: number; line: number; column: number };
        end: { offset: number; line: number; column: number };
      };
    } & { type: "attribute"; name: AttrName; children: [] });
export type AttrName = string;
export type AttrValue =
  | (Macro | FunctionMacro | TextWithoutDoubleQuote)[]
  | (Macro | FunctionMacro | TextWithoutQuote)[];
export type Text = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "text"; value: string };
export type EmptyString = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "text"; value: "" };
export type OpenParen = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "text"; value: "(" };
export type CloseParen = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "text"; value: ")" };
export type TextWithoutParenOrComma = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "text"; value: string };
export type TextWithoutParen = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "text"; value: string };
export type TextWithoutQuote = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "text"; value: string };
export type TextWithoutClosingSquareBrace = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "text"; value: string };
export type TextWithoutDoubleQuote = {
  position: {
    start: { offset: number; line: number; column: number };
    end: { offset: number; line: number; column: number };
  };
} & { type: "text"; value: string };
export type _ = string;
export type EOF = undefined;