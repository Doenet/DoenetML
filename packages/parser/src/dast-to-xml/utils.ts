import { stringifyEntitiesLight } from "stringify-entities";
import { DastNodes } from "../types";

/**
 * Escape a string.
 */
export function escape(
    value: string,
    subset: string[],
    unsafe?: RegExp | null | undefined,
): string {
    const result = clean(value);

    return unsafe ? result.replace(unsafe, encode) : encode(result);

    /**
     * Actually escape characters.
     */
    function encode(value: string): string {
        return (
            stringifyEntitiesLight(value, { subset })
                // We want fancy named versions of these two escaped characters
                .replace(/&#x3C;/g, "&lt;")
                .replace(/&#x26;/g, "&amp;")
        );
    }
}

const nonCharacter = /[\u0000-\u0008\u000B\u000C\u000E-\u001F]/g;
/**
 * Remove non-characters.
 */
export function clean(value: string) {
    return String(value || "").replace(nonCharacter, "");
}

/**
 * Encode a node name.
 */
export function name(value: string) {
    const subset = ["\t", "\n", " ", '"', "&", "'", "/", "<", "=", ">"];
    return escape(value, subset);
}

/**
 * Merge adjacent text nodes in an array
 */
export function mergeAdjacentTextInArray(nodes: DastNodes[]): DastNodes[] {
    const needsMerging = nodes.some(
        (n, i) => n.type === "text" && nodes[i + 1]?.type === "text",
    );
    if (!needsMerging) {
        return nodes;
    }
    const ret: DastNodes[] = [];
    for (let i = 0; i < nodes.length; i++) {
        let node = nodes[i];
        let nextNode = nodes[i + 1];
        if (!nextNode) {
            ret.push(node);
            continue;
        }
        if (node.type === "text" && nextNode.type === "text") {
            node = { ...node };
            ret.push(node);
            while (nextNode?.type === "text") {
                node.value += nextNode.value;
                if (node.position && nextNode.position) {
                    node.position.end = nextNode.position.end;
                }
                i++;
                nextNode = nodes[i + 1];
            }
        } else {
            ret.push(node);
        }
    }

    return ret;
}

/**
 * Recursively remove position info from a DAST tree.
 *
 * **Note**: this mutates the tree.
 */
export function filterPositionInfo(
    nodes: DastNodes | DastNodes[],
): DastNodes | DastNodes[] {
    if (typeof nodes !== "object" || nodes === null) {
        return nodes;
    }
    if (Array.isArray(nodes)) {
        return nodes.flatMap(filterPositionInfo);
    }
    if (nodes && typeof nodes === "object" && "position" in nodes) {
        delete nodes.position;
    }
    for (const value of Object.values(nodes)) {
        filterPositionInfo(value);
    }
    return nodes;
}
