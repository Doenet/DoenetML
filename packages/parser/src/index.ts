export * from "./parser";
export * from "./extract-dast-errors";
