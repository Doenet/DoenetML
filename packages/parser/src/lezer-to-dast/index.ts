export { lezerToDast } from "./lezer-to-dast";
export { lezerToDastV6 } from "./lezer-to-dast-v6";
