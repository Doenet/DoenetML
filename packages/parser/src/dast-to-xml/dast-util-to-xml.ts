// Code modified from xast-util-to-xml MIT License https://github.com/syntax-tree/xast-util-to-xml
import { ccount } from "ccount";
import {
    DastAttribute,
    DastElement,
    DastNodes,
    DastMacroFullPath,
    DastFunctionMacro,
    DastMacro,
    DastMacroPathPart,
    PrintOptions,
    DastNodesV6,
    DastMacroV6,
    DastFunctionMacroV6,
} from "../types";
import { escape, mergeAdjacentTextInArray, name } from "./utils";
import { macroToString as macroToStringV6 } from "../macros-v6/macro-to-string";

/**
 * Serialize a xast tree to XML.
 */
export function toXml(
    tree?: DastNodes[] | DastNodes | DastNodesV6[] | DastNodesV6 | null,
    options?: PrintOptions,
) {
    options = options || {};
    if (!tree) {
        return "";
    }
    const node = (
        Array.isArray(tree) ? { type: "root", children: tree } : tree
    ) as DastNodes;

    return nodesToXml(node, options);
}

/**
 * Serialize a node.
 */
export function nodesToXml(
    node: DastNodes | DastNodes[],
    options: PrintOptions,
): string {
    if (Array.isArray(node)) {
        return mergeAdjacentTextInArray(node)
            .map((child) => nodesToXml(child, options))
            .join("");
    }

    const type = node && node.type;
    switch (node?.type) {
        case "root":
            return nodesToXml(node.children, options);
        case "cdata": {
            const unsafe = /]]>/g;
            const subset = [">"];
            return "<![CDATA[" + escape(node.value, subset, unsafe) + "]]>";
        }
        case "comment":
            return "<!--" + escape(node.value, ["-"]) + "-->";
        case "doctype": {
            const nodeName = name(node.name);
            const pub = node.public;
            const sys = node.system;
            let result = "<!DOCTYPE";

            if (nodeName !== "") {
                result += " " + nodeName;
            }

            if (pub) {
                result += " PUBLIC " + quote(pub);
            } else if (sys) {
                result += " SYSTEM";
            }

            if (sys) {
                result += " " + quote(sys);
            }

            return result + ">";
        }
        case "element": {
            const nodeName = name(node.name);
            const content = nodesToXml(node.children, options);
            const attributes = node.attributes || [];

            const attrs = Object.values(attributes).map((attr) =>
                attrToString(attr, options),
            );
            const printedAttrs =
                (attrs.length > 0 ? " " : "") + attrs.join(" ");

            if (
                node.children.length === 0 ||
                (!options.inlineErrors &&
                    node.children.every((c) => c.type === "error"))
            ) {
                return `<${nodeName}${printedAttrs} />`;
            }
            return `<${nodeName}${printedAttrs}>${content}</${nodeName}>`;
        }
        case "instruction": {
            const unsafe = /\?>/g;
            const subset = [">"];
            const nodeName = name(node.name);
            const result = escape(node.value, subset, unsafe);
            return "<?" + nodeName + (result ? " " + result : "") + "?>";
        }
        case "text": {
            let escapedText = escape(node.value, ["&", "<"]);
            if (options.doenetSyntax) {
                // A < symbols that are followed by whitespace is safe
                escapedText = escapedText.replace(/&lt;(?=\s|=)/g, "<");

                // We can replace &amp; with & except in the case where it would
                // accidentally make a character entity. For example `&amp;amp;` or `&amp;#x24;`
                escapedText = escapedText.replace(/&amp;(?!\S*;)/g, "&");
            }
            return escapedText;
        }
        case "error": {
            if (options.inlineErrors) {
                const errorElement: DastElement = {
                    type: "element",
                    name: "_error",
                    attributes: {},
                    children: [{ type: "text", value: node.message }],
                };
                return nodesToXml(errorElement, options);
            }
            return "";
        }
        case "macro": {
            if ((node as unknown as DastMacroV6).version === "0.6") {
                return macroToStringV6(node as any);
            }
            const macro = unwrappedMacroToString(node, options);

            let start = "$";
            let end = "";
            if (macroNeedsParens(node)) {
                start += "(";
                end += ")";
            }
            return start + macro + end;
        }
        case "function": {
            if ((node as unknown as DastMacroV6).version === "0.6") {
                return macroToStringV6(node as any);
            }
            const macro = unwrappedMacroToString(node, options);

            let start = "$$";
            let end = "";
            if (macroNeedsParens(node)) {
                start += "(";
                end += ")";
            }
            const args = node.input
                ? `(${node.input
                      .map((a) => nodesToXml(a, options))
                      .join(", ")})`
                : "";
            return start + macro + end + args;
        }
        default: {
            // Typescript exhaustiveness check
            const unusedType: void = node;
        }
    }

    if (!type) {
        throw new Error("Expected node, not `" + node + "`");
    }

    throw new Error("Cannot compile unknown node `" + type + "`");
}

/**
 * Serialize an attribute value.
 */
export function quote(value: string) {
    const result = String(value);
    let quoteMark = '"';

    const other = quoteMark === '"' ? "'" : '"';

    if (ccount(result, quoteMark) > ccount(result, other)) {
        quoteMark = other;
    }

    return quoteMark + escape(result, ["<", quoteMark]) + quoteMark;
}

/**
 * Convert a macro to a string, but do not wrap it in parens or add a `$` prefix.
 */
function unwrappedMacroToString(
    nodes: DastMacro | DastFunctionMacro,
    options: PrintOptions,
): string {
    const path = macroPathToString(nodes.path, options);
    let attrsStr = "";
    if (nodes.type === "macro") {
        const attrs = Object.values(nodes.attributes || {})
            .map((a) => attrToString(a, options))
            .join(" ");
        attrsStr = attrs.length > 0 ? `{${attrs}}` : "";
    }
    let propAccess = "";
    return path + attrsStr + propAccess;
}

function macroPathToString(
    path: DastMacroFullPath,
    options: PrintOptions,
): string {
    return path.map((part) => macroPathPartToString(part, options)).join(".");
}

function macroPathPartToString(
    pathPart: DastMacroPathPart,
    options: PrintOptions,
): string {
    return (
        pathPart.name +
        pathPart.index
            .map((part) => `[${nodesToXml(part.value, options)}]`)
            .join("")
    );
}

function attrToString(attr: DastAttribute, options: PrintOptions): string {
    if (attr.children.length === 0) {
        // Doenet syntax allows JSX style attributes without values assigned to them
        if (options.doenetSyntax) {
            return name(attr.name);
        }
        return `${name(attr.name)}="true"`;
    }
    return `${name(attr.name)}=${quote(nodesToXml(attr.children, options))}`;
}

function macroNeedsParens(macro: DastMacro | DastFunctionMacro): boolean {
    // We also might need wrapping if the path contains a `-` character
    return macro.path.some((part) => part.name.includes("-"));
}
