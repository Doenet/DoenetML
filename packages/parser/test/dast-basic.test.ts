import { glob } from "glob";
import { describe, expect, it } from "vitest";
import * as fs from "node:fs/promises";
import { lezerToDast } from "../src/lezer-to-dast";
import { toXml } from "../src/dast-to-xml/dast-util-to-xml";
import util from "util";
import {
    filterPositionInfo,
    mergeAdjacentTextInArray,
} from "../src/dast-to-xml/utils";
import { DastRoot, DastText } from "../src/types";
import {
    splitTextAtSpecialChars,
    splitTextNodeAt,
} from "../src/lezer-to-dast/gobble-function-arguments";

const origLog = console.log;
console.log = (...args) => {
    origLog(...args.map((x) => util.inspect(x, false, 10, true)));
};

describe("DAST", async () => {
    {
        // Regular parsing tests
        const files = glob
            .sync(new URL("./fragments/xml-*.doenet", import.meta.url).pathname)
            .filter((f) => !f.includes("-bad"));
        const fileMap: Record<string, string> = Object.fromEntries(
            files.map((f) => [f.split("/").pop(), f]),
        );

        for (const [filename, filepath] of Object.entries(fileMap)) {
            it(`${filename} prints correctly in Xml format and Doenet format`, async () => {
                const source = await fs.readFile(filepath, "utf-8");
                const dast = lezerToDast(source);
                const formattedXml = toXml(dast);
                const formattedDoenet = toXml(dast, { doenetSyntax: true });
                expect(formattedXml).toMatchSnapshot();
                expect(formattedDoenet).toMatchSnapshot();
            });
        }
    }
    {
        // Tests that identify parsing errors
        const files = glob.sync(
            new URL("./fragments/xml-bad*.doenet", import.meta.url).pathname,
        );
        const fileMap: Record<string, string> = Object.fromEntries(
            files.map((f) => [f.split("/").pop(), f]),
        );

        for (const [filename, filepath] of Object.entries(fileMap)) {
            it(`${filename} can parse badly formed Xml`, async () => {
                const source = await fs.readFile(filepath, "utf-8");
                const dast = lezerToDast(source);
                // We just don't want to throw any errors when printing
                const formattedXml = toXml(dast);
                const formattedDoenet = toXml(dast, { doenetSyntax: true });
                expect(formattedXml).toBeTypeOf("string");
                expect(formattedDoenet).toBeTypeOf("string");
            });
        }
    }

    it("preserves ampersands", () => {
        const cases = [
            { inStr: "a & b", outStrXml: "a &amp; b", outStrDoenet: "a & b" },
            {
                inStr: "a &b c",
                outStrXml: "a &amp;b c",
                outStrDoenet: "a &b c",
            },
            { inStr: "a &b", outStrXml: "a &amp;b", outStrDoenet: "a &b" },
            {
                inStr: "a &amp&amp;",
                outStrXml: "a &amp;amp&amp;",
                outStrDoenet: "a &amp;amp&",
            },
            {
                inStr: "&amp&#59;",
                outStrXml: "&amp;amp;",
                outStrDoenet: "&amp;amp;",
            },
        ];
        for (const { inStr, outStrXml, outStrDoenet } of cases) {
            const dast = lezerToDast(inStr);
            const formattedXml = toXml(dast);
            const formattedDoenet = toXml(dast, { doenetSyntax: true });
            expect(formattedXml).toBe(outStrXml);
            expect(formattedDoenet).toBe(outStrDoenet);
        }
    });

    it("preserves ampersands 2", () => {
        const dast: DastRoot = {
            type: "root",
            children: [
                {
                    type: "text",
                    value: "one&two",
                },
            ],
        };
        const formattedXml = toXml(dast);
        const formattedDoenet = toXml(dast, { doenetSyntax: true });
        expect(formattedXml).toBe(`one&amp;two`);
        expect(formattedDoenet).toBe(`one&two`);
    });

    it("preserves ampersands in attributes", () => {
        const dast: DastRoot = {
            type: "root",
            children: [
                {
                    type: "element",
                    name: "a",
                    attributes: [
                        {
                            type: "attribute",
                            name: "attr",
                            children: [
                                {
                                    type: "text",
                                    value: "one&two",
                                },
                            ],
                        },
                    ],
                    children: [],
                },
                {
                    type: "text",
                    value: "\n",
                },
            ],
        };
        const formattedXml = toXml(dast).trim();
        const formattedDoenet = toXml(dast, { doenetSyntax: true }).trim();
        expect(formattedXml).toBe(`<a attr="one&amp;two" />`);
        expect(formattedDoenet).toBe(`<a attr="one&two" />`);
    });

    it("can merge adjacent text nodes", () => {
        expect(mergeAdjacentTextInArray([])).toEqual([]);
        expect(
            mergeAdjacentTextInArray([{ type: "text", value: "foo" }]),
        ).toEqual([{ type: "text", value: "foo" }]);
        expect(
            mergeAdjacentTextInArray([
                { type: "text", value: "foo" },
                { type: "text", value: "bar" },
            ]),
        ).toEqual([{ type: "text", value: "foobar" }]);
        expect(
            mergeAdjacentTextInArray([
                { type: "text", value: "foo" },
                { type: "text", value: "bar" },
                { type: "text", value: "Baz" },
            ]),
        ).toEqual([{ type: "text", value: "foobarBaz" }]);
        expect(
            mergeAdjacentTextInArray([
                { type: "text", value: "foo" },
                { type: "element", children: [], attributes: [], name: "m" },
                { type: "text", value: "bar" },
                { type: "text", value: "Foo" },
                { type: "element", children: [], attributes: [], name: "m" },
                { type: "text", value: "foo" },
            ]),
        ).toEqual([
            { type: "text", value: "foo" },
            { type: "element", children: [], attributes: [], name: "m" },
            { type: "text", value: "barFoo" },
            { type: "element", children: [], attributes: [], name: "m" },
            { type: "text", value: "foo" },
        ]);
    });
    it("Can split text nodes and preserve position information", () => {
        let source: string;
        let textNode: DastText;

        source = "hi there";
        textNode = lezerToDast(source).children[0] as DastText;
        expect(splitTextNodeAt(textNode, 2)).toMatchObject([
            {
                type: "text",
                value: "hi",
                position: {
                    start: { line: 1, column: 1, offset: 0 },
                    end: { line: 1, column: 3, offset: 2 },
                },
            },
            {
                type: "text",
                value: " ",
                position: {
                    start: { line: 1, column: 3, offset: 2 },
                    end: { line: 1, column: 4, offset: 3 },
                },
            },
            {
                type: "text",
                value: "there",
                position: {
                    start: { line: 1, column: 4, offset: 3 },
                    end: { line: 1, column: 9, offset: 8 },
                },
            },
        ]);

        source = "hi there";
        textNode = lezerToDast(source).children[0] as DastText;
        expect(splitTextNodeAt(textNode, 0)).toMatchObject([
            {
                type: "text",
                value: "",
                position: {
                    start: { line: 1, column: 1, offset: 0 },
                    end: { line: 1, column: 1, offset: 0 },
                },
            },
            {
                type: "text",
                value: "h",
                position: {
                    start: { line: 1, column: 1, offset: 0 },
                    end: { line: 1, column: 2, offset: 1 },
                },
            },
            {
                type: "text",
                value: "i there",
                position: {
                    start: { line: 1, column: 2, offset: 1 },
                    end: { line: 1, column: 9, offset: 8 },
                },
            },
        ]);

        source = "hi there\nhow are you";
        textNode = lezerToDast(source).children[0] as DastText;
        expect(splitTextNodeAt(textNode, 12)).toMatchObject([
            {
                type: "text",
                value: "hi there\nhow",
                position: {
                    start: { line: 1, column: 1, offset: 0 },
                    end: { line: 2, column: 4, offset: 12 },
                },
            },
            {
                type: "text",
                value: " ",
                position: {
                    start: { line: 2, column: 4, offset: 12 },
                    end: { line: 2, column: 5, offset: 13 },
                },
            },
            {
                type: "text",
                value: "are you",
                position: {
                    start: { line: 2, column: 5, offset: 13 },
                    end: { line: 2, column: 12, offset: 20 },
                },
            },
        ]);
    });
    it("Can split at /(),/ characters", () => {
        let source: string;
        let textNode: DastText;

        source = "hi (th, ere\n)";
        textNode = lezerToDast(source).children[0] as DastText;
        expect(
            filterPositionInfo(splitTextAtSpecialChars(textNode)),
        ).toMatchObject([
            { type: "text", value: "hi " },
            { type: "text", value: "(" },
            { type: "text", value: "th" },
            { type: "text", value: "," },
            { type: "text", value: " ere\n" },
            { type: "text", value: ")" },
        ]);
    });
    it("Produces DAST trees", () => {
        let source: string;

        source = `<m>foo</m>`;
        expect(filterPositionInfo(lezerToDast(source))).toMatchInlineSnapshot(`
          {
            "children": [
              {
                "attributes": {},
                "children": [
                  {
                    "type": "text",
                    "value": "foo",
                  },
                ],
                "name": "m",
                "type": "element",
              },
            ],
            "type": "root",
          }
        `);

        source = `<m />`;
        expect(filterPositionInfo(lezerToDast(source))).toMatchInlineSnapshot(`
          {
            "children": [
              {
                "attributes": {},
                "children": [],
                "name": "m",
                "type": "element",
              },
            ],
            "type": "root",
          }
        `);

        source = `<m foo />`;
        expect(filterPositionInfo(lezerToDast(source))).toMatchInlineSnapshot(`
          {
            "children": [
              {
                "attributes": {
                  "foo": {
                    "children": [],
                    "name": "foo",
                    "type": "attribute",
                  },
                },
                "children": [],
                "name": "m",
                "type": "element",
              },
            ],
            "type": "root",
          }
        `);

        source = `<m foo bar="baz" />`;
        expect(filterPositionInfo(lezerToDast(source))).toMatchInlineSnapshot(`
          {
            "children": [
              {
                "attributes": {
                  "bar": {
                    "children": [
                      {
                        "type": "text",
                        "value": "baz",
                      },
                    ],
                    "name": "bar",
                    "type": "attribute",
                  },
                  "foo": {
                    "children": [],
                    "name": "foo",
                    "type": "attribute",
                  },
                },
                "children": [],
                "name": "m",
                "type": "element",
              },
            ],
            "type": "root",
          }
        `);

        source = `<m><z /></m>`;
        expect(filterPositionInfo(lezerToDast(source))).toMatchInlineSnapshot(`
          {
            "children": [
              {
                "attributes": {},
                "children": [
                  {
                    "attributes": {},
                    "children": [],
                    "name": "z",
                    "type": "element",
                  },
                ],
                "name": "m",
                "type": "element",
              },
            ],
            "type": "root",
          }
        `);

        source = `<m><![CDATA[hi there]]></m>`;
        expect(filterPositionInfo(lezerToDast(source))).toMatchInlineSnapshot(`
          {
            "children": [
              {
                "attributes": {},
                "children": [
                  {
                    "type": "cdata",
                    "value": "hi there",
                  },
                ],
                "name": "m",
                "type": "element",
              },
            ],
            "type": "root",
          }
        `);

        source = `<!DOCTYPE DoenetML>\n<m>foo</m>`;
        expect(filterPositionInfo(lezerToDast(source))).toMatchInlineSnapshot(`
          {
            "children": [
              {
                "name": "DoenetML",
                "public": null,
                "system": null,
                "type": "doctype",
              },
              {
                "type": "text",
                "value": "
          ",
              },
              {
                "attributes": {},
                "children": [
                  {
                    "type": "text",
                    "value": "foo",
                  },
                ],
                "name": "m",
                "type": "element",
              },
            ],
            "type": "root",
          }
        `);
    });
});
