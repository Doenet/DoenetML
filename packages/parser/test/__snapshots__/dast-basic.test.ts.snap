// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DAST > xml-1.doenet prints correctly in Xml format and Doenet format 1`] = `"<foo />"`;

exports[`DAST > xml-1.doenet prints correctly in Xml format and Doenet format 2`] = `"<foo />"`;

exports[`DAST > xml-2.doenet prints correctly in Xml format and Doenet format 1`] = `"<foo>bar</foo>"`;

exports[`DAST > xml-2.doenet prints correctly in Xml format and Doenet format 2`] = `"<foo>bar</foo>"`;

exports[`DAST > xml-3.doenet prints correctly in Xml format and Doenet format 1`] = `"<a><b>c</b><d /></a>"`;

exports[`DAST > xml-3.doenet prints correctly in Xml format and Doenet format 2`] = `"<a><b>c</b><d /></a>"`;

exports[`DAST > xml-4.doenet prints correctly in Xml format and Doenet format 1`] = `"<a foo="bar" />"`;

exports[`DAST > xml-4.doenet prints correctly in Xml format and Doenet format 2`] = `"<a foo="bar" />"`;

exports[`DAST > xml-5.doenet prints correctly in Xml format and Doenet format 1`] = `"<a x="one" y="two" z="three" />"`;

exports[`DAST > xml-5.doenet prints correctly in Xml format and Doenet format 2`] = `"<a x="one" y="two" z="three" />"`;

exports[`DAST > xml-6.doenet prints correctly in Xml format and Doenet format 1`] = `"<a attr="one&amp;two">&amp;C</a>"`;

exports[`DAST > xml-6.doenet prints correctly in Xml format and Doenet format 2`] = `"<a attr="one&two">&C</a>"`;

exports[`DAST > xml-7.doenet prints correctly in Xml format and Doenet format 1`] = `
"<!DOCTYPE html>
<doc />"
`;

exports[`DAST > xml-7.doenet prints correctly in Xml format and Doenet format 2`] = `
"<!DOCTYPE html>
<doc />"
`;

exports[`DAST > xml-8.doenet prints correctly in Xml format and Doenet format 1`] = `"<?foo?><bar><?baz?></bar>"`;

exports[`DAST > xml-8.doenet prints correctly in Xml format and Doenet format 2`] = `"<?foo?><bar><?baz?></bar>"`;

exports[`DAST > xml-8b.doenet prints correctly in Xml format and Doenet format 1`] = `"<?foo?>"`;

exports[`DAST > xml-8b.doenet prints correctly in Xml format and Doenet format 2`] = `"<?foo?>"`;

exports[`DAST > xml-9.doenet prints correctly in Xml format and Doenet format 1`] = `
"<!--top comment -->
<element><!-- inner comment --> text</element>
<!-- c-->"
`;

exports[`DAST > xml-9.doenet prints correctly in Xml format and Doenet format 2`] = `
"<!--top comment -->
<element><!-- inner comment --> text</element>
<!-- c-->"
`;

exports[`DAST > xml-10.doenet prints correctly in Xml format and Doenet format 1`] = `"<doc><![CDATA[ hello ]]></doc>"`;

exports[`DAST > xml-10.doenet prints correctly in Xml format and Doenet format 2`] = `"<doc><![CDATA[ hello ]]></doc>"`;

exports[`DAST > xml-12.doenet prints correctly in Xml format and Doenet format 1`] = `"<xml:foo ns:bar="xx">yy</xml:foo>"`;

exports[`DAST > xml-12.doenet prints correctly in Xml format and Doenet format 2`] = `"<xml:foo ns:bar="xx">yy</xml:foo>"`;

exports[`DAST > xml-doenet-ampersand-1.doenet prints correctly in Xml format and Doenet format 1`] = `
"&amp;lt; &amp;#x25; &amp;#x25 -- this one is okay.
(an amp followed by text with a semi colon should be an amp.)"
`;

exports[`DAST > xml-doenet-ampersand-1.doenet prints correctly in Xml format and Doenet format 2`] = `
"&amp;lt; &amp;#x25; &#x25 -- this one is okay.
(an amp followed by text with a semi colon should be an amp.)"
`;

exports[`DAST > xml-doenet-ampersand-2.doenet prints correctly in Xml format and Doenet format 1`] = `
"&amp;
&amp;
&amp;amp"
`;

exports[`DAST > xml-doenet-ampersand-2.doenet prints correctly in Xml format and Doenet format 2`] = `
"&
&
&amp"
`;

exports[`DAST > xml-doenet-ampersand-3.doenet prints correctly in Xml format and Doenet format 1`] = `"&amp;"`;

exports[`DAST > xml-doenet-ampersand-3.doenet prints correctly in Xml format and Doenet format 2`] = `"&"`;

exports[`DAST > xml-doenet-attributes-1.doenet prints correctly in Xml format and Doenet format 1`] = `"<foo bar="true" baz="hi" bang="there" />"`;

exports[`DAST > xml-doenet-attributes-1.doenet prints correctly in Xml format and Doenet format 2`] = `"<foo bar baz="hi" bang="there" />"`;

exports[`DAST > xml-doenet-attributes-2.doenet prints correctly in Xml format and Doenet format 1`] = `"<foo bar='"there"' baz="&amp;hi" />"`;

exports[`DAST > xml-doenet-attributes-2.doenet prints correctly in Xml format and Doenet format 2`] = `"<foo bar='"there"' baz="&hi" />"`;

exports[`DAST > xml-doenet-lessthan-in-text-1.doenet prints correctly in Xml format and Doenet format 1`] = `"<math>2 &lt; 3</math>"`;

exports[`DAST > xml-doenet-lessthan-in-text-1.doenet prints correctly in Xml format and Doenet format 2`] = `"<math>2 < 3</math>"`;

exports[`DAST > xml-doenet-lessthan-in-text-2.doenet prints correctly in Xml format and Doenet format 1`] = `"<math>2 &lt;= 3</math>"`;

exports[`DAST > xml-doenet-lessthan-in-text-2.doenet prints correctly in Xml format and Doenet format 2`] = `"<math>2 <= 3</math>"`;

exports[`DAST > xml-doenet-lessthan-in-text-3.doenet prints correctly in Xml format and Doenet format 1`] = `
"<math>
2
&lt;
3
</math>"
`;

exports[`DAST > xml-doenet-lessthan-in-text-3.doenet prints correctly in Xml format and Doenet format 2`] = `
"<math>
2
<
3
</math>"
`;

exports[`DAST > xml-doenet-lessthan-in-text-4.doenet prints correctly in Xml format and Doenet format 1`] = `"<math> &lt; </math>"`;

exports[`DAST > xml-doenet-lessthan-in-text-4.doenet prints correctly in Xml format and Doenet format 2`] = `"<math> < </math>"`;

exports[`DAST > xml-doenet-lessthan-in-text-5.doenet prints correctly in Xml format and Doenet format 1`] = `"<math>&lt;=</math>"`;

exports[`DAST > xml-doenet-lessthan-in-text-5.doenet prints correctly in Xml format and Doenet format 2`] = `"<math><=</math>"`;

exports[`DAST > xml-doenet-no-root.doenet prints correctly in Xml format and Doenet format 1`] = `
"hi
<foo />
<bar />
there"
`;

exports[`DAST > xml-doenet-no-root.doenet prints correctly in Xml format and Doenet format 2`] = `
"hi
<foo />
<bar />
there"
`;
